### PR TITLE
Modify reprog_device::extract to return groups in a single pass

### PR DESCRIPTION
**Pull request by davidwendt**
This PR modifies the internal regex `reprog_device::extract` function to return all matching groups in a single call. Previously, retrieving each group range required individual calls to this `extract` function resulted in re-matching the entire given pattern for each group. The code logic would identify each group but only return the range for the specified group. 

The code change here passes a pre-allocated global memory array to capture each group range in a single pass. The extract is an all-or-nothing process. In fact, a `find` function must first be executed to retrieve the bounds of the given pattern. So if any of the groups are missing or do not match, no groups are returned for that row. Retrieving the last group would always require processing the previous groups and the code logic now records those positions in the global memory array. The memory array can then be used directly to build the output columns.

This simplifies the code around extract and also improves performance especially for long strings or patterns with many groups. For small strings and a small number of groups, the gbenchmark showed equivalent performance to the previous implementation. For larger strings and more groups, the gbenchmark showed a 2-3x improvement.

### CODE REVIEW

**Review by vuule — CHANGES_REQUESTED** [review]
Looks good, got a few mostly stylistic suggestions.
Didn't dig too deep into the algorithmic aspect of the changes, let me know if it's needed.

File: cpp/src/strings/regex/regex.cuh [hunk 1]
```diff
@@ -157,21 +164,23 @@ class reprog_device {
...
-   * @param group_id The specific instance to return if more than one match is found.
-   * @return Returns 0 if no match is found.
+   * @param indices All extracted groups
+   * @return Returns true if successful.
    */
   template <int stack_size>
-  __device__ inline int32_t extract(
-    int32_t idx, string_view const& d_str, int32_t& begin, int32_t& end, int32_t group_id);
+  __device__ inline bool extract(
+    int32_t idx, string_view const& d_str, int32_t begin, int32_t end, string_index_pair* indices);
```
[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/strings/regex/regex.cuh:183]: Thanks Vukasin. I'd rather keep this particular API as a simple data pointer for now. But I think using the `device_2dspan` and `device_span` in the calling functions makes sense.

File: cpp/src/strings/extract.cu [hunk 2]
```diff
@@ -42,26 +42,27 @@ namespace {
...
+      int32_t begin     = 0;
+      int32_t end       = -1;  // handles empty strings automatically
+      if ((prog.find<stack_size>(idx, d_str, begin, end) > 0) &&
+          prog.extract<stack_size>(idx, d_str, begin, end, d_output)) {
+        return;
+      }
     }
-    return result;
+    // fill output with null entries
+    for (int i = 0; i < groups; ++i) { d_output[i] = string_index_pair{nullptr, 0}; }
```
[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/strings/extract.cu:65]: I can use a `thrust::fill` here.

File: cpp/src/strings/regex/regex.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/strings/regex/regex.cuh]: After adding the `device_2dspan` changing this interface to accept a `device_span` was fairly trivial.

**Review by vuule — APPROVED** [review]

**Review by robertmaynard — APPROVED** [review]
CMake changes LGTM

**Review by cwharris — APPROVED** [review]

**Comment by vuule** [comment]
@gpucibot merge

**Pull request merged**